system/lib/wasmfs/file_table.h

```diff
@@ -0,0 +1,101 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+// This file defines the open file table of the new file system.
+// Current Status: Work in Progress.
+// See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#pragma once
+
+#include "file.h"
+#include <assert.h>
+#include <mutex>
+#include <utility>
+#include <vector>
+#include <wasi/api.h>
+
+class OpenFileDescriptor {
+  std::shared_ptr<File> file;
+  __wasi_filedelta_t offset;
+  std::mutex mutex;
+
+public:
+  OpenFileDescriptor(uint32_t offset, std::shared_ptr<File> file);
+
+  std::shared_ptr<File>& getFile();
+};
+
+class FileTable {
+  static std::vector<std::shared_ptr<OpenFileDescriptor>> entries;
+  std::mutex mutex;
+
+  FileTable() {
+    entries.push_back(std::make_shared<OpenFileDescriptor>(0, std::make_shared<StdinFile>()));
+    entries.push_back(std::make_shared<OpenFileDescriptor>(0, std::make_shared<StdoutFile>()));
+    entries.push_back(std::make_shared<OpenFileDescriptor>(0, std::make_shared<StderrFile>()));
+  }
+
+public:
+  // Handle represents an RAII wrapper object. Access to the FileTable must go through a Handle.
+  // A Handle holds FileTable's lock for the duration of its lifetime.
+  class Handle {
+    FileTable& fileTable;
+    std::unique_lock<std::mutex> lock;
+
+  public:
+    Handle(FileTable& fileTable) : fileTable(fileTable), lock(fileTable.mutex) {}
+
+    // Returns size of the associated FileTable entries.
+    size_t size() { return fileTable.entries.size(); }
+
+    // Adds given OpenFileDescriptor to FileTable entries. Returns fd (insertion index in entries).
+    __wasi_fd_t addOpenFile(std::shared_ptr<OpenFileDescriptor> ptr);
+
+    // Removes OpenFileDescriptor in FileTable entries corresponding to given fd.
+    void removeOpenFile(__wasi_fd_t fd);
+
+    // Entry is used to override the subscript [] operator.
+    // This allows the user to get and set values in the FileTable entries vector.
+    // Syntax: std::shared_ptr<OpenFileDescriptor> openFile = currentHandle[fd];
```

Comment:
It would be good to further explain what `Entry` can do that direct access to the underlying `entries` cannot do. I also worry that this usage example is more confusing than helpful because there are plenty of ways to use `Entry` that don't involve immediately converting it to `std::shared_ptr<OpenFileDescriptor>`.
```suggestion
    // Entry is used to override the subscript [] operator.
    // This allows the user to get and set values in the FileTable entries vector, even entries outside the bounds of the FileTable.
```